### ISSUE TITLE
Enable wildcards (eg: imfile) at directory level

### ISSUE THREAD

**Issue by mostolog** [rsyslog/rsyslog#1154]
Hi

According to: http://www.rsyslog.com/doc/v8-stable/configuration/modules/imfile.html#input-parameters and http://www.slideshare.net/rainergerhards1/using-wildcards-with-rsyslogs-file-monitor-imfile
**wildcards are only valid at filename level, but not at directory level**.

Is there any reason for this limitation (Other than "I haven't enough time to code it" ?

Regards

**Comment by rgerhards** [comment]
sorry for the sluggish replies these days; I am out of office with limited Internet access. Situation will improve next week.

To the question: "not enough time to code it" is for sure a restriction. But this operation can also be very memory hungry and performance intensen. You need to create a notify watch on each directory level. Consider this to be monitored:
```
/*/*.log
```
Even what does that mean? All *.log files everywhere. Or just on the second level of the hierarchy? Even then it could be very costly.

Note that while I find this topic interesting, I am not be able to give any more comments this weeks. Sorry for that.

**Comment by mostolog** [comment]
Don't worry @rgerhards.

IMHO: **/\*/\*.log** should monitor only *.log files on 2nd level hierarchy.
```
/.                        <----- watch for directory changes
/..                       <----- It won't be processed
/foo/a.log                <---read
/foo/.                    <---- watch for .log file changes
/foo/foo/a.log            <---- it shouldn't be readed
/bar/b.log                <---read
/bar/.                    <---- watch for .log file changes
/bar/bar/b.log            <----- don't read it neither
```
In my case, I was trying to handle reading something like **/logs/apache/appX/access.log** (obviously, X changes) using **/logs/apache/*/access.log**

On the other hand, you will need a _recursive_ parameter if you want to go deeper through all hierarchy and process all log files, which -IMHO- shouldn't be a common/reasonable use case.

Anyhow, I must agree we need more discussion about this.

**Comment by chenryn** [comment]
logstash use /**/*.log  for recursive mean, and /*/*.log for only one level
directory.

2016-09-26 18:13 GMT+08:00 mostolog <notifications@github.com>:

**Comment by mostolog** [comment]
@chenryn Didn't know that.
That seems a great approach!

**Comment by mostolog** [comment]
Hi


I have been re-inventing the wheel and this is what I got:

```
struct watcher_t {
	watcher_t **watcher_childs;
};
watcher_t root_watcher;

char *mydir = "full path filename w/wo wildcards" (configured on imfile)

char *base = mydir_substring(from:0, to:last_/_before_any_*);
watchDir(root_watcher, base, mydir_substring(from:last_/_before_any_*, to:'\0');

int watchDir(watcher_t *parent, char *base, char *dir) {
	//possible dir values
	*|\0		newWatcher = inotify_add_watch($base, files)
				parent->add_child_watcher(newWatcher)
				For each file: 
							newFileWatcher = inotify_add_watch($file)
							newWatcher->add_child_watcher($newFileWatcher)

	*/...		newWatcher = inotify_add_watch($base, dirs)
				parent->add_child_watcher(newWatcher)
				For each new_dir: watchDir($newWatcher, $base/$new_dir, ...)

	**			newWatcher = inotify_add_watch($base, dirs|files)
				parent->add_child_watcher(newWatcher)
				For each file: 
							newFileWatcher = inotify_add_watch($file)
							newWatcher->add_child_watcher($newFileWatcher)
				For each new_dir: watchDir($newWatcher, $base/$new_dir, **)

	**/...		newWatcher = inotify_add_watch($base, dirs|files)
				parent->add_child_watcher(newWatcher)
				For each file: 
							newFileWatcher = inotify_add_watch($file)
							newWatcher->add_child_watcher($newFileWatcher)
				For each new_dir: watchDir($newWatcher, $base/$new_dir, **/...)

	foo			newWatcher = inotify_add_watch($base, foo_files)
				parent->add_child_watcher(newWatcher)
				For each file: 
							newFileWatcher = inotify_add_watch($file)
							newWatcher->add_child_watcher($newFileWatcher)

	foo*		newWatcher = inotify_add_watch($base, foo*_files)
				parent->add_child_watcher(newWatcher)
				For each file: 
							newFileWatcher = inotify_add_watch($file)
							newWatcher->add_child_watcher($newFileWatcher)

	foo*/...	newWatcher = inotify_add_watch($base, foo*_dirs)
				parent->add_child_watcher(newWatcher)
				For each new_dir: watchDir($newWatcher, $base/$new_dir, ...)

	foo**		newWatcher = inotify_add_watch($base, foo*_dirs|foo*_files)
				parent->add_child_watcher(newWatcher)
				For each file: 
							newFileWatcher = inotify_add_watch($file)
							newWatcher->add_child_watcher($newFileWatcher)
				For each new_dir: watchDir($newWatcher, $base/$new_dir, **)

	foo**/...	newWatcher = inotify_add_watch($base, foo*_dirs|foo*_files)
				parent->add_child_watcher(newWatcher)
				For each file: 
							newFileWatcher = inotify_add_watch($file)
							newWatcher->add_child_watcher($newFileWatcher)
				For each new_dir: watchDir($newWatcher, $base/$new_dir, **/...)
}
```

Hope it's clear enough. Please, give me some feedback

If anyone has a simpler/better idea, it will be more than welcome!

Regards

**Comment by chjohnst** [comment]
This would be a useful feature to me.  I have systems that bring apps up/down throughout the course of a systems uptime and all of the logs are under a specific path with a directory for their app.  So I have a simple glob catching all of the apps debug logs using 8.25 latest features.  When rsyslog starts up it catches all of the logs and I can forward them out, but if a new app comes along under a new directory they don't get picked up without restarting rsyslog which is not good in a dynamic environment. 

For now I think my way of handling this is to restart rsyslog at intervals to pick up the new changes.  Since imfile uses a statefile for the current logs its handling and I have queue setup, in theory the restarts shouldn't be a problem even if we get backlogged on the restart.  

input(type="imfile"
      File="/logs/*/log.DEBUG"
      Tag="ptss:")

**Comment by akmalabbasov** [comment]
having the same issue as @chjohnst , are there any plans to add support for this scenario?

**Comment by rgerhards** [comment]
This functionality was implemented in 8.25.0. Unfortunately, it looks like the issue tracker was not closed. I am doing this now. 

If you have an issue with the functionality, please open a new issue tracker. **However** there is a known issue for which a patch already exists: https://github.com/rsyslog/rsyslog/pull/1455 So it would probably be good if you could apply that patch before reporting an issue. I admit that from what I read, it doesn't look to be the same problem, but maybe...
